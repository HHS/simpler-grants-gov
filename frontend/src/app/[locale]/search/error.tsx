"use client";

import QueryProvider from "src/app/[locale]/search/QueryProvider";
import { usePrevious } from "src/hooks/usePrevious";
import { FrontendErrorDetails } from "src/types/apiResponseTypes";
import { OptionalStringDict } from "src/types/generalTypes";
import { Breakpoints, ErrorProps } from "src/types/uiTypes";
import { convertSearchParamsToProperTypes } from "src/utils/search/convertSearchParamsToProperTypes";

import { useTranslations } from "next-intl";
import { ReadonlyURLSearchParams, useSearchParams } from "next/navigation";
import { useEffect } from "react";
import { Alert } from "@trussworks/react-uswds";

import ContentDisplayToggle from "src/components/ContentDisplayToggle";
import SearchBar from "src/components/search/SearchBar";
import SearchFilters from "src/components/search/SearchFilters";
import ServerErrorAlert from "src/components/ServerErrorAlert";

export interface ParsedError {
  message: string;
  searchInputs: OptionalStringDict;
  status: number;
  type: string;
  details?: FrontendErrorDetails;
}

function isValidJSON(str: string) {
  try {
    JSON.parse(str);
    return true;
  } catch (e) {
    return false; // String is not valid JSON
  }
}

function createBlankParsedError(): ParsedError {
  return {
    type: "NetworkError",
    searchInputs: {
      query: "",
      status: "",
      fundingInstrument: "",
      eligibility: "",
      agency: "",
      category: "",
      sortby: undefined,
      page: "1",
      actionType: "initialLoad",
    },
    message: "Invalid error message JSON returned",
    status: -1,
  };
}

export default function SearchError({ error, reset }: ErrorProps) {
  const t = useTranslations("Search");
  const searchParams = useSearchParams();
  const previousSearchParams =
    usePrevious<ReadonlyURLSearchParams>(searchParams);

  useEffect(() => {
    if (
      reset &&
      previousSearchParams &&
      searchParams.toString() !== previousSearchParams?.toString()
    ) {
      reset();
    }
  }, [searchParams, reset, previousSearchParams]);

  useEffect(() => {
    console.error(error);
  }, [error]);

  // The error message is passed as an object that's been stringified.
  // Parse it here.
  let parsedErrorData;

  if (!isValidJSON(error.message)) {
    // the error likely is just a string with a non-specific Server Component error when running the built app
    // "An error occurred in the Server Components render. The specific message is omitted in production builds..."
    parsedErrorData = createBlankParsedError();
  } else {
    // Valid error thrown from server component
    parsedErrorData = JSON.parse(error.message) as ParsedError;
  }
  const convertedSearchParams = convertSearchParamsToProperTypes(
    parsedErrorData.searchInputs,
  );
  const { agency, category, eligibility, fundingInstrument, query, status } =
    convertedSearchParams;

  // note that the validation error will contain untranslated strings
  const ErrorAlert =
    parsedErrorData.details && parsedErrorData.type === "ValidationError" ? (
      <Alert type="error" heading={t("validationError")} headingLevel="h4">
        {`Error in ${parsedErrorData.details.field || "a search field"}: ${parsedErrorData.details.message || "adjust your search and try again"}`}
      </Alert>
    ) : (
      <ServerErrorAlert callToAction={t("generic_error_cta")} />
    );

  return (
    <QueryProvider>
      <div className="grid-container">
        <div className="search-bar">
          <SearchBar query={query} />
        </div>
        <div className="grid-row grid-gap">
          <div className="tablet:grid-col-4">
            <ContentDisplayToggle
              showCallToAction={t("filterDisplayToggle.showFilters")}
              hideCallToAction={t("filterDisplayToggle.hideFilters")}
              breakpoint={Breakpoints.TABLET}
            >
              <SearchFilters
                opportunityStatus={status}
                eligibility={eligibility}
                category={category}
                fundingInstrument={fundingInstrument}
                agency={agency}
              />
            </ContentDisplayToggle>
          </div>
          <div className="tablet:grid-col-8">{ErrorAlert}</div>
        </div>
      </div>
    </QueryProvider>
  );
}
