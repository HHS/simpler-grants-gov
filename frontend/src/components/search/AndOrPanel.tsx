"use client";

import { useSearchParamUpdater } from "src/hooks/useSearchParamUpdater";

import { useCallback } from "react";

import { andOrOptions } from "./SearchFilterAccordion/SearchFilterOptions";
import { SearchFilterRadio } from "./SearchFilterRadio";

export function AndOrPanel({
  hasSearchTerm,
  andOrParam,
}: {
  hasSearchTerm: boolean;
  andOrParam: Set<string>;
}) {
  const { setStaticQueryParam, setQueryParam } = useSearchParamUpdater();

  const toggleAndOrSelection = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      if (hasSearchTerm) {
        return setQueryParam("andOr", event.target.value);
      }
      setStaticQueryParam("andOr", event.target.value);
    },
    [hasSearchTerm],
  );

  return andOrOptions.map((option) => {
    return (
      <SearchFilterRadio
        id={option.id}
        name={option.value}
        label={option.label}
        onChange={toggleAndOrSelection}
        value={option.value}
        checked={
          option.value === "and"
            ? !andOrParam.size || andOrParam.has(option.value)
            : andOrParam.has(option.value)
        }
      />
    );
  });

  // return (
  //   <div>
  //     <SearchFilterRadio
  //       id="andOr-"
  //       name={`Any ${title}`}
  //       label={`Any ${title.toLowerCase()}`}
  //       onChange={toggleAndOrSelection}
  //       value={undefined}
  //       facetCount={undefined}
  //       checked={isNoneSelected}
  //     />
  //     <SearchFilterRadio />
  //   </div>
  // );
}
