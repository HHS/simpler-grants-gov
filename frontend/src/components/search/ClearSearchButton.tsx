"use client";

import { useSearchParamUpdater } from "src/hooks/useSearchParamUpdater";

import { Button } from "@trussworks/react-uswds";

import { USWDSIcon } from "src/components/USWDSIcon";

export function ClearSearchButton({
  buttonText,
  includeIcon = false,
  className,
  paramsToClear,
}: {
  buttonText: string;
  includeIcon?: boolean;
  className?: string;
  paramsToClear?: string[];
}) {
  const { clearQueryParams } = useSearchParamUpdater();
  return (
    <Button
      unstyled
      type="button"
      onClick={() => clearQueryParams(paramsToClear)}
      className={className}
    >
      {includeIcon && (
        <USWDSIcon
          name="close"
          className="margin-right-05 margin-left-neg-05"
        />
      )}
      {buttonText}
    </Button>
  );
}
