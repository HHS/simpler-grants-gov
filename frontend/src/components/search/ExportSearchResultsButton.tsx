"use client";

import { downloadSearchResultsCSV } from "src/services/fetch/fetchers/clientSearchResultsDownloadFetcher";

import { useTranslations } from "next-intl";
import { useSearchParams } from "next/navigation";
import { useCallback } from "react";
import { Button } from "@trussworks/react-uswds";

import { USWDSIcon } from "src/components/USWDSIcon";

export function ExportSearchResultsButton() {
  const t = useTranslations("Search.exportButton");
  const searchParams = useSearchParams();

  const downloadSearchResults = useCallback(() => {
    // catch included here to satisfy linter
    downloadSearchResultsCSV(searchParams).catch((e) => {
      throw e;
    });
  }, [searchParams]);

  return (
    <div
      className="desktop:grid-col-4 desktop:display-flex flex-align-self-center"
      data-testid="search-download-button-container"
    >
      <Button
        outline={true}
        type={"submit"}
        className="width-auto margin-top-2 tablet:width-100 tablet-lg:margin-top-0"
        onClick={downloadSearchResults}
      >
        <USWDSIcon name="file_download" className="usa-icon--size-3" />
        {t("title")}
      </Button>
    </div>
  );
}
