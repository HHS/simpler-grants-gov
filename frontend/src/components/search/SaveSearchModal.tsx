import clsx from "clsx";
import { useClientFetch } from "src/hooks/useClientFetch";
import { useIsSSR } from "src/hooks/useIsSSR";
import { useUser } from "src/services/auth/useUser";
import { filterSearchParams } from "src/utils/search/searchFormatUtils";

import { useTranslations } from "next-intl";
import { useSearchParams } from "next/navigation";
import { RefObject, useCallback, useRef, useState } from "react";
import {
  Button,
  ErrorMessage,
  FormGroup,
  Modal,
  ModalFooter,
  ModalHeading,
  ModalRef,
  ModalToggleButton,
  TextInput,
} from "@trussworks/react-uswds";

import { LoadingButton } from "src/components/LoadingButton";
import SimplerAlert from "src/components/SimplerAlert";
import { USWDSIcon } from "src/components/USWDSIcon";

function SaveSearchInput({
  validationError,
  updateSavedSearchName,
}: {
  validationError?: string;
  updateSavedSearchName: (name: string) => void;
}) {
  const t = useTranslations("Search.saveSearch.modal");
  const inputRef = useRef<HTMLInputElement>(null);

  return (
    <FormGroup error={!!validationError}>
      <label htmlFor="saved-search-input">
        {t.rich("inputLabel", {
          required: (chunks) => (
            <span className="usa-hint usa-hint--required">{chunks}</span>
          ),
        })}
      </label>
      {validationError && <ErrorMessage>{validationError}</ErrorMessage>}
      <div className="usa-search usa-search--big" role="search">
        <TextInput
          ref={inputRef}
          className={clsx("usa-input", "maxw-none", {
            "usa-input--error": !!validationError,
          })}
          id="saved-search-input"
          name="save-search"
          defaultValue={""}
          onChange={(e) => updateSavedSearchName(e.target?.value)}
          type="text"
          required
          aria-required
        />
      </div>
    </FormGroup>
  );
}

function SuccessContent({
  modalRef,
  modalId,
  onClose,
}: {
  modalRef: RefObject<ModalRef | null>;
  modalId: string;
  onClose: () => void;
}) {
  const t = useTranslations("Search.saveSearch.modal");
  return (
    <>
      <ModalHeading id={`${modalId}-heading`}>{t("successTitle")}</ModalHeading>
      <div className="usa-prose">
        <p className="font-sans-2xs margin-y-4">
          {t.rich("successDescription", {
            workspaceLink: (chunks) => (
              <a href="/saved-search-queries">{chunks}</a>
            ),
          })}
        </p>
      </div>
      <ModalFooter>
        <ModalToggleButton
          modalRef={modalRef}
          closer
          unstyled
          className="padding-105 text-center"
          onClick={onClose}
        >
          {t("closeText")}
        </ModalToggleButton>
      </ModalFooter>
    </>
  );
}

// note that this is client component, but not marked as "use client" because to do so
// would cause an error when passing the function prop from parent client component see:
// https://github.com/vercel/next.js/discussions/46795
export function SaveSearchModal({ onSave }: { onSave: (id: string) => void }) {
  const modalId = "save-search";

  const t = useTranslations("Search.saveSearch.modal");
  const modalRef = useRef<ModalRef>(null);
  const { user } = useUser();
  const searchParams = useSearchParams();
  const isSSR = useIsSSR();

  const [validationError, setValidationError] = useState<string>();
  const [savedSearchName, setSavedSearchName] = useState<string>();
  const [apiError, setApiError] = useState<boolean>();
  const [loading, setLoading] = useState<boolean>();
  const [saved, setSaved] = useState<boolean>();
  const { clientFetch } = useClientFetch<{ id: string }>(
    "Error posting saved search",
  );
  const handleSubmit = useCallback(() => {
    if (validationError) {
      setValidationError(undefined);
    }
    if (!savedSearchName) {
      setValidationError(t("emptyNameError"));
      return;
    }
    if (!user?.token) {
      return;
    }
    // send up a filtered set of params, converted to an object
    // we will do the further filter and pagination object building on the server
    const savedSearchParams = filterSearchParams(
      Object.fromEntries(searchParams.entries()),
    );

    setLoading(true);
    clientFetch("/api/user/saved-searches", {
      method: "POST",
      body: JSON.stringify({ ...savedSearchParams, name: savedSearchName }),
    })
      .then((data) => {
        if (!data?.id) {
          throw new Error("saved search ID not returned from API");
        }
        setSaved(true);
        onSave(data.id);
      })
      .catch((error) => {
        setApiError(true);
        console.error(error);
      })
      .finally(() => {
        setLoading(false);
      });
  }, [
    savedSearchName,
    user?.token,
    searchParams,
    t,
    validationError,
    onSave,
    clientFetch,
  ]);

  const onClose = useCallback(() => {
    setSaved(false);
    setApiError(false);
    setLoading(false);
    setValidationError(undefined);
    setSavedSearchName("");
  }, []);

  return (
    <>
      <div className="flex-1">
        <ModalToggleButton
          modalRef={modalRef}
          opener
          data-testid="open-save-search-modal-button"
          type="button"
          className="padding-1 hover:bg-base-lightest"
          unstyled
        >
          <USWDSIcon
            className="usa-icon margin-right-05 margin-left-neg-05"
            name="add_circle_outline"
            key="save-search"
          />
          {t("saveText")}
        </ModalToggleButton>
      </div>
      <Modal
        ref={modalRef}
        forceAction
        className="text-wrap"
        aria-labelledby={`${modalId}-heading`}
        aria-describedby={`${modalId}-description`}
        id={modalId}
        renderToPortal={!isSSR}
        onKeyDown={(e) => {
          if (e.key === "Enter") handleSubmit();
        }}
      >
        {saved ? (
          <SuccessContent
            modalRef={modalRef}
            modalId={modalId}
            onClose={onClose}
          />
        ) : (
          <>
            <ModalHeading id={`${modalId}-heading`}>{t("title")}</ModalHeading>
            <div className="usa-prose">
              <p className="font-sans-2xs margin-y-4">{t("description")}</p>
            </div>
            {apiError && (
              <SimplerAlert
                alertClick={() => setApiError(false)}
                buttonId="saveSearchApiError"
                messageText={t("apiError")}
                type="error"
              />
            )}
            <SaveSearchInput
              validationError={validationError}
              updateSavedSearchName={setSavedSearchName}
            />
            <ModalFooter>
              {loading ? (
                <LoadingButton id="save-search-button" message={t("loading")} />
              ) : (
                <>
                  <Button
                    type="button"
                    onClick={handleSubmit}
                    data-testid="save-search-button"
                  >
                    {t("saveText")}
                  </Button>
                  <ModalToggleButton
                    modalRef={modalRef}
                    closer
                    unstyled
                    className="padding-105 text-center"
                    onClick={onClose}
                  >
                    {t("cancelText")}
                  </ModalToggleButton>
                </>
              )}
            </ModalFooter>
          </>
        )}
      </Modal>
    </>
  );
}
