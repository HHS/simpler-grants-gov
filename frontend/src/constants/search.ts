import { FrontendFilterNames } from "src/types/search/searchFilterTypes";

// Show all values status for search.
export const SEARCH_NO_STATUS_VALUE = "none";
export const SEARCH_DEFAULT_VALUES = ["forecasted", "posted"];
export const defaultFilterValues: { [key in FrontendFilterNames]?: string[] } =
  {
    status: SEARCH_DEFAULT_VALUES,
  };
