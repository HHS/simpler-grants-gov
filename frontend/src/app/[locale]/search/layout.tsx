import { SEARCH_CRUMBS } from "src/constants/breadcrumbs";
import { LayoutProps } from "src/types/generalTypes";

import { useTranslations } from "next-intl";
import { setRequestLocale } from "next-intl/server";
import { use } from "react";

import BetaAlert from "src/components/BetaAlert";
import Breadcrumbs from "src/components/Breadcrumbs";
import SearchCallToAction from "src/components/search/SearchCallToAction";

export default function SearchLayout({ children, params }: LayoutProps) {
  const { locale } = use(params);
  setRequestLocale(locale);

  const t = useTranslations("Search.beta_alert");

  return (
    <>
      <BetaAlert
        containerClasses="margin-top-5"
        heading={t("alert_title")}
        alertMessage={t.rich("alert", {
          mailToGrants: (chunks) => (
            <a href="mailto:simpler@grants.gov">{chunks}</a>
          ),
          bugReport: (chunks) => (
            <a href="https://github.com/HHS/simpler-grants-gov/issues/new?template=1_bug_report.yml">
              {chunks}
            </a>
          ),
          featureRequest: (chunks) => (
            <a href="https://github.com/HHS/simpler-grants-gov/issues/new?template=2_feature_request.yml">
              {chunks}
            </a>
          ),
        })}
      />
      <Breadcrumbs breadcrumbList={SEARCH_CRUMBS} />
      <SearchCallToAction />
      {children}
    </>
  );
}
