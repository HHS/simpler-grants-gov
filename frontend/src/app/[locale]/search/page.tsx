import { Metadata } from "next";
import QueryProvider from "src/app/[locale]/search/QueryProvider";
import { environment } from "src/constants/environments";
import withFeatureFlag from "src/hoc/withFeatureFlag";
import { LocalizedPageProps } from "src/types/intl";
import { SearchParamsTypes } from "src/types/search/searchRequestTypes";
import { Breakpoints } from "src/types/uiTypes";
import { convertSearchParamsToProperTypes } from "src/utils/search/convertSearchParamsToProperTypes";

import { useTranslations } from "next-intl";
import { getTranslations, setRequestLocale } from "next-intl/server";
import { redirect } from "next/navigation";
import { use } from "react";

import ContentDisplayToggle from "src/components/ContentDisplayToggle";
import { SaveSearchPanel } from "src/components/search/SaveSearchPanel";
import SearchAnalytics from "src/components/search/SearchAnalytics";
import SearchBar from "src/components/search/SearchBar";
import SearchFilters from "src/components/search/SearchFilters";
import SearchResults from "src/components/search/SearchResults";

export async function generateMetadata({ params }: LocalizedPageProps) {
  const { locale } = await params;
  const t = await getTranslations({ locale });
  const meta: Metadata = {
    title: t("Search.title"),
    description: t("Search.meta_description"),
  };
  return meta;
}
type SearchPageProps = {
  searchParams: Promise<SearchParamsTypes>;
  params: Promise<{ locale: string }>;
};

function Search({ searchParams, params }: SearchPageProps) {
  const { locale } = use(params);
  const resolvedSearchParams = use(searchParams);
  setRequestLocale(locale);
  const t = useTranslations("Search");

  const convertedSearchParams =
    convertSearchParamsToProperTypes(resolvedSearchParams);
  const { agency, category, eligibility, fundingInstrument, query, status } =
    convertedSearchParams;

  if (!("page" in resolvedSearchParams)) {
    resolvedSearchParams.page = "1";
  }

  return (
    <>
      <SearchAnalytics
        params={resolvedSearchParams}
        newRelicEnabled={environment.NEW_RELIC_ENABLED === "true"}
      />
      <QueryProvider>
        <div className="grid-container">
          <div className="search-bar">
            <SearchBar query={query} />
          </div>
          <div className="grid-row grid-gap">
            <div className="tablet:grid-col-4">
              <ContentDisplayToggle
                showCallToAction={t("filterDisplayToggle.showFilters")}
                hideCallToAction={t("filterDisplayToggle.hideFilters")}
                breakpoint={Breakpoints.TABLET}
                type="centered"
              >
                <SaveSearchPanel />
                <SearchFilters
                  opportunityStatus={status}
                  eligibility={eligibility}
                  category={category}
                  fundingInstrument={fundingInstrument}
                  agency={agency}
                />
              </ContentDisplayToggle>
            </div>
            <div className="tablet:grid-col-8">
              <SearchResults
                searchParams={convertedSearchParams}
                query={query}
                loadingMessage={t("loading")}
              ></SearchResults>
            </div>
          </div>
        </div>
      </QueryProvider>
    </>
  );
}

// Exports page behind both feature flags
export default withFeatureFlag<SearchPageProps, never>(
  Search,
  "searchOff",
  () => redirect("/maintenance"),
);
