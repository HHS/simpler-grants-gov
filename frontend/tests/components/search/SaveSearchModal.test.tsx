import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { useTranslationsMock } from "src/utils/testing/intlMocks";

import { SaveSearchModal } from "src/components/search/SaveSearchModal";

const mockUseUser = jest.fn(() => ({
  user: {
    token: "faketoken",
  },
}));

const mockSaveSearch = jest.fn();

jest.mock("src/services/auth/useUser", () => ({
  useUser: () => mockUseUser(),
}));
jest.mock("src/services/fetch/fetchers/clientSavedSearchFetcher", () => ({
  saveSearch: (...args) => mockSaveSearch(...args),
}));

jest.mock("next-intl", () => ({
  useTranslations: () => useTranslationsMock(),
}));

jest.useFakeTimers();

describe("SaveSearchModal", () => {
  afterEach(() => {
    mockSaveSearch.mockReset();
    jest.clearAllTimers();
  });
  it("displays a working modal toggle button", async () => {
    const { rerender } = render(<SaveSearchModal />);

    expect(screen.queryByRole("dialog")).toHaveClass("is-hidden");

    const toggle = await screen.findByTestId("open-save-search-modal-button");
    toggle.click();

    rerender(<SaveSearchModal />);

    expect(screen.getByRole("dialog")).not.toHaveClass("is-hidden");
  });
  it("modal can be closed as expected", async () => {
    const { rerender } = render(<SaveSearchModal />);

    const toggle = await screen.findByTestId("open-save-search-modal-button");
    toggle.click();

    rerender(<SaveSearchModal />);

    const closeButton = await screen.findByText("cancelText");
    closeButton.click();

    rerender(<SaveSearchModal />);

    expect(screen.queryByRole("dialog")).toHaveClass("is-hidden");
  });
  it("displays validation error if submitted without a name", async () => {
    const { rerender } = render(<SaveSearchModal />);

    const toggle = await screen.findByTestId("open-save-search-modal-button");
    toggle.click();

    rerender(<SaveSearchModal />);

    const saveButton = await screen.findByTestId("save-search-button");
    saveButton.click();

    rerender(<SaveSearchModal />);

    const validationError = await screen.findByText("emptyNameError");

    expect(validationError).toBeInTheDocument();
  });
  it("displays an API error if API returns an error", async () => {
    mockSaveSearch.mockRejectedValue(new Error());
    const { rerender } = render(<SaveSearchModal />);

    const toggle = await screen.findByTestId("open-save-search-modal-button");
    toggle.click();

    rerender(<SaveSearchModal />);

    const saveButton = await screen.findByTestId("save-search-button");
    const input = await screen.findByTestId("textInput");
    fireEvent.change(input, { target: { value: "save search name" } });
    saveButton.click();

    rerender(<SaveSearchModal />);

    // this isn't translated... update after entering real text
    const error = await screen.findByText(
      "Failed to save search. Please try again.",
    );

    expect(error).toBeInTheDocument();
  });
  it("displays a success message on successful save", async () => {
    mockSaveSearch.mockResolvedValue({ arbitrary: "data" });
    const { rerender } = render(<SaveSearchModal />);

    const toggle = await screen.findByTestId("open-save-search-modal-button");
    toggle.click();

    rerender(<SaveSearchModal />);

    const saveButton = await screen.findByTestId("save-search-button");
    const input = await screen.findByTestId("textInput");
    fireEvent.change(input, { target: { value: "save search name" } });
    saveButton.click();

    rerender(<SaveSearchModal />);

    // this isn't translated... update after entering real text
    const success = await screen.findByText("successDescription");

    expect(success).toBeInTheDocument();
  });
});
