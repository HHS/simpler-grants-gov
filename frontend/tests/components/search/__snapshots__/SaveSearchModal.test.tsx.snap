// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaveSearchModal matches snapshot 1`] = `
<div>
  <div
    class="flex-1"
  >
    <button
      aria-controls="save-search"
      class="usa-button usa-button--unstyled padding-1 hover:bg-base-lightest"
      data-open-modal="true"
      data-testid="open-save-search-modal-button"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="usa-icon usa-icon margin-right-05 margin-left-neg-05"
        role="img"
      >
        <use
          href="/img.jpg#add_circle_outline"
        />
      </svg>
      saveText
    </button>
  </div>
</div>
`;
